### PR TITLE
feat: enable tcp base protocol listen on same port

### PR DESCRIPTION
**Pull request by driftluo**
In the original implementation, TCP/WS/TLS could not be listened on the same port. After this PR, these protocols can be listened on the same port.

### CODE REVIEW

File: tentacle/src/transports/tcp_base_listen.rs [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+use crate::transports::ws::WsStream;
+#[cfg(feature = "tls")]
+use crate::{service::TlsConfig, transports::parse_tls_domain_name};
+#[cfg(feature = "tls")]
+use std::borrow::Cow;
+#[cfg(feature = "tls")]
+use tokio_rustls::{
+    rustls::{server::ResolvesServerCertUsingSni, ServerConfig},
+    TlsAcceptor,
+};
```
Comment (eval-exec) [inline tentacle/src/transports/tcp_base_listen.rs:33]:
```suggestion
#[cfg(feature = "ws")]
use {
    crate::transports::ws::WsStream,
    tokio_tungstenite::accept_async,
};

#[cfg(feature = "tls")]
use {
    crate::{service::TlsConfig, transports::parse_tls_domain_name},
    std::borrow::Cow,
    tokio_rustls::{
        rustls::{server::ResolvesServerCertUsingSni, ServerConfig},
        TlsAcceptor,
    },
};
```

File: tentacle/src/transports/tcp_base_listen.rs [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    Upgrade,
+    New(TcpBaseListener),
+}
+
+/// Tcp listen bind
+pub async fn bind(
+    address: impl Future<Output = Result<Multiaddr>>,
+    tcp_config: TcpSocketConfig,
+    self_mode: UpgradeMode,
+    #[cfg(feature = "tls")] config: TlsConfig,
```
Comment (eval-exec) [inline tentacle/src/transports/tcp_base_listen.rs:55]: new line

File: tentacle/src/transports/tcp_base_listen.rs [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+                    match parse_tls_domain_name(&addr) {
+                        None => return Err(TransportErrorKind::NotSupported(addr)),
+                        Some(d) => {
+                            tcp_base_addr.push(Protocol::Tls(Cow::Owned(d)));
+                        }
+                    }
+                    tcp_base_addr
+                }
+                #[allow(unreachable_patterns)]
+                _ => unreachable!(),
```
Comment (eval-exec) [inline tentacle/src/transports/tcp_base_listen.rs:116]: Is is possible that `self_mode.to_enum()` is `UpgradeModeEnum::All`?

File: tentacle/src/utils/dns.rs [hunk 4]
```diff
@@ -76,7 +76,11 @@ impl DnsResolver {
             Some(address) => {
                 let mut address = socketaddr_to_multiaddr(address);
                 match self.ty {
-                    TransportType::Tcp | TransportType::Memory | TransportType::Tls => (),
+                    TransportType::Tcp | TransportType::Memory => (),
+                    TransportType::Tls => match parse_tls_domain_name(&self.source_address) {
+                        Some(domain_name) => address.push(Protocol::Tls(Cow::Owned(domain_name))),
+                        None => address.push(Protocol::Tls(Cow::Borrowed(""))),
+                    },
```
Comment (eval-exec) [inline tentacle/src/utils/dns.rs:83]: How about:
```rust
TransportType::Tls => {
    let domain_name = parse_tls_domain_name(&self.source_address)
        .map(Cow::Owned)
        .unwrap_or_else(|| Cow::Borrowed(""));

    address.push(Protocol::Tls(domain_name));
}
```

File: tentacle/src/transports/tcp_base_listen.rs (hunk 3 above)
Comment (driftluo) [reply tentacle/src/transports/tcp_base_listen.rs:116]: Impossible, that each listen operation can only specify one protocol; to support multiple protocols, multiple listen operations are required.

**Review by eval-exec — APPROVED** [review]

File: tentacle/src/transports/mod.rs [hunk 5]
```diff
@@ -93,24 +95,36 @@ pub fn find_type(addr: &Multiaddr) -> TransportType {
     .unwrap_or(TransportType::Tcp)
 }
 
+pub(crate) fn parse_tls_domain_name(addr: &Multiaddr) -> Option<String> {
+    let mut iter = addr.iter();
```
Comment (eval-exec) [inline tentacle/src/transports/mod.rs:99]: I think mut is unnecessary.

Comment (driftluo) [reply tentacle/src/transports/mod.rs:99]: `find_map` need it

File: tentacle/src/transports/tcp_base_listen.rs [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+                }
+                return;
+            }
+            #[cfg(feature = "tls")]
+            UpgradeModeEnum::TcpAndTLS => {
+                let mut peek_buf = [0u8; 16];
+                if let Err(e) = stream.peek(&mut peek_buf).await {
+                    debug!("stream encountered err: {}, close unexpectedly", e);
+                    return;
+                }
```
Comment (eval-exec) [inline tentacle/src/transports/tcp_base_listen.rs:410]: https://docs.rs/tokio/latest/tokio/net/struct.TcpStream.html#method.peek

Do we need to be concerned about the number of bytes peeked when using `stream.peek`? I'm worried that `stream.peek(&mut peek_buf)` might not fill the entire `peek_buf`.

Comment (driftluo) [reply tentacle/src/transports/tcp_base_listen.rs:410]: I will add this check

File: tentacle/src/transports/tcp_base_listen.rs [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+                let molecule_header =
+                    u32::from_le_bytes(TryInto::<[u8; 4]>::try_into(&peek_buf[4..8]).unwrap());
+                let rand_start =
+                    u32::from_le_bytes(TryInto::<[u8; 4]>::try_into(&peek_buf[8..12]).unwrap());
+                let rand_end =
+                    u32::from_le_bytes(TryInto::<[u8; 4]>::try_into(&peek_buf[12..16]).unwrap());
+
+                if length_delimited_header == molecule_header
+                    && rand_start == 24
+                    && rand_start < rand_end
```
Comment (eval-exec) [inline tentacle/src/transports/tcp_base_listen.rs:433]: Do we need to check the length of `rand_end - rand_start`?

Comment (driftluo) [reply tentacle/src/transports/tcp_base_listen.rs:433]: we can't, secio does not specify the size of rand

**Review by eval-exec — APPROVED** [review]

File: tentacle/tests/tcp_upgrade_mod.rs [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                .unwrap();
+            service.run().await
+        });
+    });
+
+    let handle_2 = thread::spawn(move || {
+        let _multi_addr_2 = Multiaddr::from_str(
+            "/ip4/127.0.0.1/tcp/0/tls/0x388f042dd011824b91ecda56c85eeec993894f88",
+        )
+        .unwrap();
```
Comment (quake) [inline tentacle/tests/tcp_upgrade_mod.rs:161]: unnecessary code?

File: tentacle/src/transports/tcp_base_listen.rs [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+                            debug!("accept tls server stream err: {:?}", err);
+                        }
+                    },
+                }
+                return;
+            }
+            #[cfg(feature = "tls")]
+            UpgradeModeEnum::TcpAndTls => {
+                let mut peek_buf = [0u8; 16];
+                loop {
```
[the remark this answers is not in the document]
Comment (driftluo) [reply tentacle/src/transports/tcp_base_listen.rs]: Now we change to a unified process, start reading 16 bytes uniformly, return directly if timeout, and then judge the protocol according to the branch

File: tentacle/tests/tcp_upgrade_mod.rs (hunk 8 above)
Comment (driftluo) [reply tentacle/tests/tcp_upgrade_mod.rs]: delete it now

**Pull request merged**